// @ethsign/sp-sdk/build/cjs/index.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __exportStar = (this && this.__exportStar) || function(m, exports) {
    for (var p in m) if (p !== "default" && !Object.prototype.hasOwnProperty.call(exports, p)) __createBinding(exports, m, p);
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.getStorageData = exports.getOffChainStorageData = exports.saveDataToStorage = exports.encodeOnChainData = exports.decodeOnChainData = exports.IndexService = exports.chainInfo = exports.delegateSignSchema = exports.delegateSignRevokeAttestation = exports.delegateSignAttestation = exports.DataLocationOffChain = exports.DataLocationOnChain = exports.OffChainSignType = exports.OffChainRpc = exports.SpMode = exports.EvmChains = exports.SignProtocolClient = void 0;
const IndexService_1 = require("./IndexService");
Object.defineProperty(exports, "IndexService", { enumerable: true, get: function () { return IndexService_1.IndexService; } });
const SignProtocolClient_1 = require("./SignProtocolClient");
Object.defineProperty(exports, "SignProtocolClient", { enumerable: true, get: function () { return SignProtocolClient_1.SignProtocolClient; } });
const constants_1 = require("./clients/evm/constants");
Object.defineProperty(exports, "chainInfo", { enumerable: true, get: function () { return constants_1.chainInfo; } });
const types_1 = require("./clients/evm/types");
Object.defineProperty(exports, "EvmChains", { enumerable: true, get: function () { return types_1.EvmChains; } });
const types_2 = require("./types");
Object.defineProperty(exports, "SpMode", { enumerable: true, get: function () { return types_2.SpMode; } });
Object.defineProperty(exports, "OffChainSignType", { enumerable: true, get: function () { return types_2.OffChainSignType; } });
Object.defineProperty(exports, "DataLocationOnChain", { enumerable: true, get: function () { return types_2.DataLocationOnChain; } });
Object.defineProperty(exports, "DataLocationOffChain", { enumerable: true, get: function () { return types_2.DataLocationOffChain; } });
const offChain_1 = require("./types/offChain");
Object.defineProperty(exports, "OffChainRpc", { enumerable: true, get: function () { return offChain_1.OffChainRpc; } });
const tools_1 = require("./utils/tools");
Object.defineProperty(exports, "delegateSignAttestation", { enumerable: true, get: function () { return tools_1.delegateSignAttestation; } });
Object.defineProperty(exports, "delegateSignRevokeAttestation", { enumerable: true, get: function () { return tools_1.delegateSignRevokeAttestation; } });
Object.defineProperty(exports, "delegateSignSchema", { enumerable: true, get: function () { return tools_1.delegateSignSchema; } });
const utils_1 = require("./utils");
Object.defineProperty(exports, "decodeOnChainData", { enumerable: true, get: function () { return utils_1.decodeOnChainData; } });
Object.defineProperty(exports, "encodeOnChainData", { enumerable: true, get: function () { return utils_1.encodeOnChainData; } });
Object.defineProperty(exports, "getStorageData", { enumerable: true, get: function () { return utils_1.getStorageData; } });
const services_1 = require("./services");
Object.defineProperty(exports, "getOffChainStorageData", { enumerable: true, get: function () { return services_1.getOffChainStorageData; } });
Object.defineProperty(exports, "saveDataToStorage", { enumerable: true, get: function () { return services_1.saveDataToStorage; } });
__exportStar(require("./types"), exports);
