// @ethsign/sp-sdk/build/cjs/types/index.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __exportStar = (this && this.__exportStar) || function(m, exports) {
    for (var p in m) if (p !== "default" && !Object.prototype.hasOwnProperty.call(exports, p)) __createBinding(exports, m, p);
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.RecipientEncodingType = exports.SpMode = exports.ChainType = exports.OffChainSignType = exports.OnChainSignType = exports.DataLocationOffChain = exports.DataLocationOnChain = void 0;
__exportStar(require("./indexService"), exports);
var DataLocationOnChain;
(function (DataLocationOnChain) {
    DataLocationOnChain[DataLocationOnChain["ONCHAIN"] = 0] = "ONCHAIN";
    DataLocationOnChain[DataLocationOnChain["ARWEAVE"] = 1] = "ARWEAVE";
    DataLocationOnChain[DataLocationOnChain["IPFS"] = 2] = "IPFS";
})(DataLocationOnChain || (exports.DataLocationOnChain = DataLocationOnChain = {}));
var DataLocationOffChain;
(function (DataLocationOffChain) {
    DataLocationOffChain["ARWEAVE"] = "arweave";
    DataLocationOffChain["IPFS"] = "ipfs";
    DataLocationOffChain["GREENFIELD"] = "greenfield";
    DataLocationOffChain["GREENFIELD_TESTTNET"] = "greenfield-testnet";
})(DataLocationOffChain || (exports.DataLocationOffChain = DataLocationOffChain = {}));
var OnChainSignType;
(function (OnChainSignType) {
    OnChainSignType["eip712"] = "eip712";
})(OnChainSignType || (exports.OnChainSignType = OnChainSignType = {}));
var OffChainSignType;
(function (OffChainSignType) {
    OffChainSignType["EvmEip712"] = "evm-eip712";
    OffChainSignType["Solana"] = "solana";
    OffChainSignType["TonConnect"] = "ton-connect";
})(OffChainSignType || (exports.OffChainSignType = OffChainSignType = {}));
var ChainType;
(function (ChainType) {
    ChainType["evm"] = "evm";
    ChainType["solana"] = "solana";
    ChainType["ton"] = "ton";
})(ChainType || (exports.ChainType = ChainType = {}));
var SpMode;
(function (SpMode) {
    SpMode["OnChain"] = "OnChain";
    SpMode["OffChain"] = "OffChain";
})(SpMode || (exports.SpMode = SpMode = {}));
var RecipientEncodingType;
(function (RecipientEncodingType) {
    RecipientEncodingType["Address"] = "address";
    RecipientEncodingType["String"] = "string";
})(RecipientEncodingType || (exports.RecipientEncodingType = RecipientEncodingType = {}));
