// @ethsign/sp-sdk/build/cjs/clients/evm/constants.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.chainInfo = exports.ContractInfoMap = void 0;
const chains_1 = require("viem/chains");
const types_1 = require("./types");
const types_2 = require("../../types");
exports.ContractInfoMap = {
    [types_1.EvmChains.mainnet]: {
        address: '0x3D8E699Db14d7781557fE94ad99d93Be180A6594',
        chain: chains_1.mainnet,
    },
    [types_1.EvmChains.bsc]: {
        address: '0xe2C15B97F628B7Ad279D6b002cEDd414390b6D63',
        chain: chains_1.bsc,
    },
    [types_1.EvmChains.zetachainAthensTestnet]: {
        address: '0x5d4D4eEd224028C230aFEbB69d279DE99bC06338',
        chain: chains_1.zetachainAthensTestnet,
    },
    [types_1.EvmChains.opBNBTestnet]: {
        address: '0x72efA4093539A909C1f9bcCA1aE6bcDa435a3433',
        chain: chains_1.opBNBTestnet,
    },
    [types_1.EvmChains.zetachainMainnet]: {
        address: '0xBbc279ee396074aC968b459d542DEE60c6bD71C1',
        chain: chains_1.zetachain,
    },
    [types_1.EvmChains.polygon]: {
        address: '0xe2C15B97F628B7Ad279D6b002cEDd414390b6D63',
        chain: chains_1.polygon,
    },
    [types_1.EvmChains.opBNB]: {
        address: '0x03688D459F172B058d39241456Ae213FC4E26941',
        chain: chains_1.opBNB,
    },
    [types_1.EvmChains.scrollSepolia]: {
        address: '0x4e4af2a21ebf62850fD99Eb6253E1eFBb56098cD',
        chain: chains_1.scrollSepolia,
    },
    [types_1.EvmChains.scroll]: {
        address: '0xFBF614E89Ac79d738BaeF81CE6929897594b7E69',
        chain: chains_1.scroll,
    },
    [types_1.EvmChains.base]: {
        address: '0x2b3224D080452276a76690341e5Cfa81A945a985',
        chain: chains_1.base,
    },
    [types_1.EvmChains.xLayer]: {
        address: '0x4e4af2a21ebf62850fD99Eb6253E1eFBb56098cD',
        chain: chains_1.xLayer,
    },
    [types_1.EvmChains.plumeTestnet]: {
        address: '0x4e4af2a21ebf62850fD99Eb6253E1eFBb56098cD',
        chain: chains_1.plumeTestnet,
    },
    [types_1.EvmChains.berachainTestnet]: {
        address: '0x2774d96a841E522549CE7ADd3825fC31075384Cf',
        chain: chains_1.berachainTestnet,
    },
    [types_1.EvmChains.baseSepolia]: {
        address: '0x4e4af2a21ebf62850fD99Eb6253E1eFBb56098cD',
        chain: chains_1.baseSepolia,
    },
    [types_1.EvmChains.sepolia]: {
        address: '0x878c92FD89d8E0B93Dc0a3c907A2adc7577e39c5',
        chain: chains_1.sepolia,
    },
    [types_1.EvmChains.polygonAmoy]: {
        address: '0x4e4af2a21ebf62850fD99Eb6253E1eFBb56098cD',
        chain: chains_1.polygonAmoy,
    },
    [types_1.EvmChains.optimism]: {
        address: '0x945C44803E92a3495C32be951052a62E45A5D964',
        chain: chains_1.optimism,
    },
    [types_1.EvmChains.gnosisChiado]: {
        address: '0x4e4af2a21ebf62850fD99Eb6253E1eFBb56098cD',
        chain: chains_1.gnosisChiado,
    },
    [types_1.EvmChains.optimismSepolia]: {
        address: '0x4e4af2a21ebf62850fD99Eb6253E1eFBb56098cD',
        chain: chains_1.optimismSepolia,
    },
    [types_1.EvmChains.arbitrumSepolia]: {
        address: '0x4e4af2a21ebf62850fD99Eb6253E1eFBb56098cD',
        chain: chains_1.arbitrumSepolia,
    },
    [types_1.EvmChains.polygonMumbai]: {
        address: '0x4665fffdD8b48aDF5bab3621F835C831f0ee36D7',
        chain: chains_1.polygonMumbai,
    },
    [types_1.EvmChains.gnosis]: {
        address: '0x4e4af2a21ebf62850fD99Eb6253E1eFBb56098cD',
        chain: chains_1.gnosis,
    },
    [types_1.EvmChains.degen]: {
        address: '0x4e4af2a21ebf62850fD99Eb6253E1eFBb56098cD',
        chain: chains_1.degen,
    },
    [types_1.EvmChains.cyber]: {
        address: '0x4e4af2a21ebf62850fD99Eb6253E1eFBb56098cD',
        chain: chains_1.cyber,
    },
    [types_1.EvmChains.arbitrum]: {
        address: '0x4e4af2a21ebf62850fD99Eb6253E1eFBb56098cD',
        chain: chains_1.arbitrum,
    },
    [types_1.EvmChains.celo]: {
        address: '0x4e4af2a21ebf62850fD99Eb6253E1eFBb56098cD',
        chain: chains_1.celo,
    },
    [types_1.EvmChains.celoAlfajores]: {
        address: '0x4e4af2a21ebf62850fD99Eb6253E1eFBb56098cD',
        chain: chains_1.celoAlfajores,
    },
    [types_1.EvmChains.plume]: {
        address: '0x4e4af2a21ebf62850fD99Eb6253E1eFBb56098cD',
        chain: chains_1.plume,
    },
    [types_1.EvmChains.monadTestnet]: {
        address: '0x4e4af2a21ebf62850fD99Eb6253E1eFBb56098cD',
        chain: chains_1.monadTestnet,
    },
};
exports.chainInfo = {
    evm: [
        {
            name: chains_1.mainnet.name,
            chain: types_1.EvmChains.mainnet,
            icon: 'https://sign-public-cdn.s3.us-east-1.amazonaws.com/sp-sdk/icons/ETH_240205060739_240723091257.svg',
            scanUrl: chains_1.mainnet.blockExplorers.default.url + '/tx/',
            id: chains_1.mainnet.id + '',
            chainInfo: chains_1.mainnet,
        },
        {
            name: chains_1.bsc.name,
            chain: types_1.EvmChains.bsc,
            icon: 'https://sign-public-cdn.s3.us-east-1.amazonaws.com/sp-sdk/icons/bnb_240709085015.webp',
            scanUrl: chains_1.bsc.blockExplorers.default.url + '/tx/',
            id: chains_1.bsc.id + '',
            chainInfo: chains_1.bsc,
        },
        {
            name: chains_1.polygon.name,
            chain: types_1.EvmChains.polygon,
            icon: 'https://sign-public-cdn.s3.us-east-1.amazonaws.com/Attestation-App/polygon-token_240726025025.svg',
            scanUrl: chains_1.polygon.blockExplorers.default.url + '/tx/',
            id: chains_1.polygon.id + '',
            chainInfo: chains_1.polygon,
        },
        {
            name: chains_1.zetachain.name,
            chain: types_1.EvmChains.zetachainMainnet,
            icon: 'https://sign-public-cdn.s3.us-east-1.amazonaws.com/sp-sdk/icons/zeta_240123034413_240723091203.svg',
            scanUrl: chains_1.zetachain.blockExplorers.default.url + '/tx/',
            id: chains_1.zetachain.id + '',
            chainInfo: chains_1.zetachain,
        },
        {
            name: chains_1.opBNB.name,
            chain: types_1.EvmChains.opBNB,
            icon: 'https://sign-public-cdn.s3.us-east-1.amazonaws.com/sp-sdk/icons/bnb_240127141716_240723091203.svg',
            scanUrl: chains_1.opBNB.blockExplorers.default.url + '/tx/',
            id: chains_1.opBNB.id + '',
            chainInfo: chains_1.opBNB,
        },
        {
            name: chains_1.scroll.name,
            chain: types_1.EvmChains.scroll,
            icon: 'https://sign-public-cdn.s3.us-east-1.amazonaws.com/sp-sdk/icons/Scroll_Logomark_240723092217.svg',
            scanUrl: chains_1.scroll.blockExplorers.default.url + '/tx/',
            id: chains_1.scroll.id + '',
            chainInfo: chains_1.scroll,
        },
        {
            name: chains_1.base.name,
            chain: types_1.EvmChains.base,
            icon: 'https://sign-public-cdn.s3.us-east-1.amazonaws.com/sp-sdk/icons/base_240321024616_240723091203.svg',
            scanUrl: chains_1.base.blockExplorers.default.url + '/tx/',
            id: chains_1.base.id + '',
            chainInfo: chains_1.base,
        },
        {
            name: chains_1.optimism.name,
            chain: types_1.EvmChains.optimism,
            icon: 'https://sign-public-cdn.s3.us-east-1.amazonaws.com/sp-sdk/icons/op_240415064952_240723091203.svg',
            scanUrl: chains_1.optimism.blockExplorers.default.url + '/tx/',
            id: chains_1.optimism.id + '',
            chainInfo: chains_1.optimism,
        },
        {
            name: chains_1.gnosis.name,
            chain: types_1.EvmChains.gnosis,
            icon: 'https://sign-public-cdn.s3.us-east-1.amazonaws.com/sp-sdk/icons/gnosis_240723093122.svg',
            scanUrl: chains_1.gnosis.blockExplorers.default.url + '/tx/',
            id: chains_1.gnosis.id + '',
            chainInfo: chains_1.gnosis,
        },
        {
            name: chains_1.degen.name,
            chain: types_1.EvmChains.degen,
            icon: 'https://sign-public-cdn.s3.us-east-1.amazonaws.com/sp-sdk/icons/degen_240723093839.svg',
            scanUrl: chains_1.degen.blockExplorers.default.url + '/tx/',
            id: chains_1.degen.id + '',
            chainInfo: chains_1.degen,
        },
        {
            name: chains_1.xLayer.name,
            chain: types_1.EvmChains.xLayer,
            icon: 'https://sign-public-cdn.s3.us-east-1.amazonaws.com/sp-sdk/icons/3D95DB39A3F6B46D_240724010255.webp',
            scanUrl: chains_1.xLayer.blockExplorers.default.url + '/tx/',
            id: chains_1.xLayer.id + '',
            chainInfo: chains_1.xLayer,
        },
        {
            name: chains_1.cyber.name,
            chain: types_1.EvmChains.cyber,
            icon: 'https://sign-public-cdn.s3.us-east-1.amazonaws.com/sp-sdk/icons/cyber-coin_240724010558.webp',
            scanUrl: chains_1.cyber.blockExplorers.default.url + '/tx/',
            id: chains_1.cyber.id + '',
            chainInfo: chains_1.cyber,
        },
        {
            name: chains_1.arbitrum.name,
            chain: types_1.EvmChains.arbitrum,
            icon: 'https://sign-public-cdn.s3.us-east-1.amazonaws.com/Attestation-App/arbitrum-one_240726024522.svg',
            scanUrl: chains_1.arbitrum.blockExplorers.default.url + '/tx/',
            id: chains_1.arbitrum.id + '',
            chainInfo: chains_1.arbitrum,
        },
        {
            name: chains_1.polygonAmoy.name,
            chain: types_1.EvmChains.polygonAmoy,
            icon: 'https://sign-public-cdn.s3.us-east-1.amazonaws.com/Attestation-App/polygon-token_240726025025.svg',
            scanUrl: chains_1.polygonAmoy.blockExplorers.default.url + '/tx/',
            id: chains_1.polygonAmoy.id + '',
            chainInfo: chains_1.polygonAmoy,
            isTestnet: true,
        },
        {
            name: chains_1.gnosisChiado.name,
            chain: types_1.EvmChains.gnosisChiado,
            icon: 'https://sign-public-cdn.s3.us-east-1.amazonaws.com/sp-sdk/icons/gnosis_240723093122.svg',
            scanUrl: chains_1.gnosisChiado.blockExplorers.default.url + '/tx/',
            id: chains_1.gnosisChiado.id + '',
            chainInfo: chains_1.gnosisChiado,
            isTestnet: true,
        },
        {
            name: chains_1.zetachainAthensTestnet.name,
            chain: types_1.EvmChains.zetachainAthensTestnet,
            icon: 'https://sign-public-cdn.s3.us-east-1.amazonaws.com/sp-sdk/icons/zeta_240123034413_240723091203.svg',
            scanUrl: chains_1.zetachainAthensTestnet.blockExplorers.default.url + '/tx/',
            id: chains_1.zetachainAthensTestnet.id + '',
            isTestnet: true,
            chainInfo: chains_1.zetachainAthensTestnet,
        },
        {
            name: chains_1.opBNBTestnet.name,
            chain: types_1.EvmChains.opBNBTestnet,
            icon: 'https://sign-public-cdn.s3.us-east-1.amazonaws.com/sp-sdk/icons/bnb_240127141716_240723091203.svg',
            scanUrl: chains_1.opBNBTestnet.blockExplorers.default.url + '/tx/',
            id: chains_1.opBNBTestnet.id + '',
            isTestnet: true,
            chainInfo: chains_1.opBNBTestnet,
        },
        {
            name: chains_1.scrollSepolia.name,
            chain: types_1.EvmChains.scrollSepolia,
            icon: 'https://sign-public-cdn.s3.us-east-1.amazonaws.com/sp-sdk/icons/Scroll_Logomark_240723092217.svg',
            scanUrl: chains_1.scrollSepolia.blockExplorers.default.url + '/tx/',
            id: chains_1.scrollSepolia.id + '',
            isTestnet: true,
            chainInfo: chains_1.scrollSepolia,
        },
        {
            name: chains_1.baseSepolia.name,
            chain: types_1.EvmChains.baseSepolia,
            icon: 'https://sign-public-cdn.s3.us-east-1.amazonaws.com/sp-sdk/icons/base_240321024616_240723091203.svg',
            scanUrl: chains_1.baseSepolia.blockExplorers.default.url + '/tx/',
            id: chains_1.baseSepolia.id + '',
            isTestnet: true,
            chainInfo: chains_1.baseSepolia,
        },
        {
            name: chains_1.celo.name,
            chain: types_1.EvmChains.celo,
            icon: 'https://sign-public-cdn.s3.us-east-1.amazonaws.com/sp-sdk/icons/celo_240813062718.svg',
            scanUrl: chains_1.celo.blockExplorers.default.url + '/tx/',
            id: chains_1.celo.id + '',
            chainInfo: chains_1.celo,
        },
        {
            name: chains_1.plume.name,
            chain: types_1.EvmChains.plume,
            icon: 'https://sign-public-cdn.s3.us-east-1.amazonaws.com/sp-sdk/icons/plume_240724011259.webp',
            scanUrl: chains_1.plume.blockExplorers.default.url + '/tx/',
            id: chains_1.plume.id + '',
            chainInfo: chains_1.plume,
        },
        {
            name: chains_1.plumeTestnet.name,
            chain: types_1.EvmChains.plumeTestnet,
            icon: 'https://sign-public-cdn.s3.us-east-1.amazonaws.com/sp-sdk/icons/plume_240724011259.webp',
            scanUrl: chains_1.plumeTestnet.blockExplorers.default.url + '/tx/',
            id: chains_1.plumeTestnet.id + '',
            isTestnet: true,
            chainInfo: chains_1.plumeTestnet,
        },
        {
            name: chains_1.berachainTestnet.name,
            chain: types_1.EvmChains.berachainTestnet,
            icon: 'https://sign-public-cdn.s3.us-east-1.amazonaws.com/sp-sdk/icons/Berachain_Orange_240724011628.webp',
            scanUrl: chains_1.berachainTestnet.blockExplorers.default.url + '/tx/',
            id: chains_1.berachainTestnet.id + '',
            isTestnet: true,
            chainInfo: chains_1.berachainTestnet,
        },
        {
            name: chains_1.sepolia.name,
            chain: types_1.EvmChains.sepolia,
            icon: 'https://sign-public-cdn.s3.us-east-1.amazonaws.com/sp-sdk/icons/ETH_240205060739_240723091257.svg',
            scanUrl: chains_1.sepolia.blockExplorers.default.url + '/tx/',
            id: chains_1.sepolia.id + '',
            isTestnet: true,
            chainInfo: chains_1.sepolia,
        },
        {
            name: chains_1.optimismSepolia.name,
            chain: types_1.EvmChains.optimismSepolia,
            icon: 'https://sign-public-cdn.s3.us-east-1.amazonaws.com/sp-sdk/icons/op_240415064952_240723091203.svg',
            scanUrl: chains_1.optimismSepolia.blockExplorers.default.url + '/tx/',
            id: chains_1.optimismSepolia.id + '',
            isTestnet: true,
            chainInfo: chains_1.optimismSepolia,
        },
        {
            name: chains_1.arbitrumSepolia.name,
            chain: types_1.EvmChains.arbitrumSepolia,
            icon: 'https://sign-public-cdn.s3.us-east-1.amazonaws.com/Attestation-App/arbitrum-one_240726024522.svg',
            scanUrl: chains_1.arbitrumSepolia.blockExplorers.default.url + '/tx/',
            id: chains_1.arbitrumSepolia.id + '',
            isTestnet: true,
            chainInfo: chains_1.arbitrumSepolia,
        },
        {
            name: chains_1.polygonMumbai.name,
            chain: types_1.EvmChains.polygonMumbai,
            icon: 'https://sign-public-cdn.s3.us-east-1.amazonaws.com/Attestation-App/polygon-token_240726025025.svg',
            scanUrl: chains_1.polygonMumbai.blockExplorers.default.url + '/tx/',
            id: chains_1.polygonMumbai.id + '',
            isTestnet: true,
            chainInfo: chains_1.polygonMumbai,
        },
        {
            name: chains_1.celoAlfajores.name,
            chain: types_1.EvmChains.celoAlfajores,
            icon: 'https://sign-public-cdn.s3.us-east-1.amazonaws.com/sp-sdk/icons/celo_240813062718.svg',
            scanUrl: chains_1.celoAlfajores.blockExplorers.default.url + '/tx/',
            id: chains_1.celoAlfajores.id + '',
            isTestnet: true,
            chainInfo: chains_1.celoAlfajores,
        },
        {
            name: chains_1.monadTestnet.name,
            chain: types_1.EvmChains.monadTestnet,
            icon: 'https://sign-public-cdn.s3.us-east-1.amazonaws.com/sp-sdk/icons/Monad_250331022304.svg',
            scanUrl: chains_1.monadTestnet.blockExplorers.default.url + '/tx/',
            id: chains_1.monadTestnet.id + '',
            isTestnet: true,
            chainInfo: chains_1.monadTestnet,
        },
    ],
    offchain: [
        {
            chain: types_2.DataLocationOffChain.ARWEAVE,
            id: types_2.DataLocationOffChain.ARWEAVE,
            icon: 'https://sign-public-cdn.s3.us-east-1.amazonaws.com/sp-sdk/icons/arweave_240724011931.svg',
            name: 'Arweave',
            scanUrl: 'https://viewblock.io/arweave/tx/',
        },
        {
            chain: types_2.DataLocationOffChain.IPFS,
            id: types_2.DataLocationOffChain.IPFS,
            icon: 'https://sign-public-cdn.s3.us-east-1.amazonaws.com/sp-sdk/icons/Ipfs_240724012252.webp',
            name: 'IPFS',
            scanUrl: 'https://ipfs.io/ipfs/',
        },
        {
            chain: types_2.DataLocationOffChain.GREENFIELD,
            id: types_2.DataLocationOffChain.GREENFIELD,
            icon: 'https://sign-public-cdn.s3.us-east-1.amazonaws.com/sp-sdk/icons/bnb_240127141716_240723091203.svg',
            name: 'Greenfield',
            scanUrl: 'https://greenfieldscan.com/tx/',
        },
        {
            chain: types_2.DataLocationOffChain.GREENFIELD_TESTTNET,
            id: types_2.DataLocationOffChain.GREENFIELD_TESTTNET,
            icon: 'https://sign-public-cdn.s3.us-east-1.amazonaws.com/sp-sdk/icons/bnb_240127141716_240723091203.svg',
            name: 'Greenfield',
            scanUrl: 'https://testnet.greenfieldscan.com/tx/',
            isTestnet: true,
        },
    ],
};
