// @ethsign/sp-sdk/build/esm/clients/evm/types.js
export var EvmChains;
(function (EvmChains) {
    EvmChains["mainnet"] = "mainnet";
    EvmChains["gnosis"] = "gnosis";
    EvmChains["degen"] = "degen";
    EvmChains["polygon"] = "polygon";
    EvmChains["opBNB"] = "opBNB";
    EvmChains["scroll"] = "scroll";
    EvmChains["base"] = "base";
    EvmChains["optimism"] = "optimism";
    EvmChains["xLayer"] = "xLayer";
    EvmChains["zetachainAthensTestnet"] = "zetachainAthensTestnet";
    EvmChains["opBNBTestnet"] = "opBNBTestnet";
    EvmChains["zetachainMainnet"] = "zetachainMainnet";
    EvmChains["scrollSepolia"] = "scrollSepolia";
    EvmChains["plumeTestnet"] = "plumeTestnet";
    EvmChains["berachainTestnet"] = "berachainTestnet";
    EvmChains["baseSepolia"] = "baseSepolia";
    EvmChains["sepolia"] = "sepolia";
    EvmChains["polygonAmoy"] = "polygonAmoy";
    EvmChains["gnosisChiado"] = "gnosisChiado";
    EvmChains["optimismSepolia"] = "optimismSepolia";
    EvmChains["arbitrumSepolia"] = "arbitrumSepolia";
    EvmChains["polygonMumbai"] = "polygonMumbai";
    EvmChains["cyber"] = "cyber";
    EvmChains["arbitrum"] = "arbitrum";
    EvmChains["bsc"] = "bsc";
    EvmChains["celo"] = "celo";
    EvmChains["celoAlfajores"] = "celoAlfajores";
    EvmChains["plume"] = "plume";
    EvmChains["monadTestnet"] = "monadTestnet";
})(EvmChains || (EvmChains = {}));
