// @ethsign/sp-sdk/build/cjs/clients/evm/OffChain/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.EVMOffChainClient = void 0;
const types_1 = require("../../../types");
const OffChainClientBase_1 = require("../../../interface/OffChainClientBase");
const viem_1 = require("viem");
const chains_1 = require("viem/chains");
const utils_1 = require("../../../utils");
class EVMOffChainClient extends OffChainClientBase_1.OffChainClientBase {
    walletClient;
    privateKeyAccount;
    account;
    constructor(options) {
        super({
            chainType: types_1.ChainType.evm,
            signType: options.signType,
            rpcUrl: options.rpcUrl,
            apiKey: options.apiKey,
        });
        if ('walletClient' in options) {
            this.walletClient = options.walletClient;
        }
        else if (options.account) {
            this.privateKeyAccount = options.account;
        }
        else {
            this.walletClient = (0, viem_1.createWalletClient)({
                chain: chains_1.mainnet,
                transport: (0, utils_1.isBrowser)() && window.ethereum ? (0, viem_1.custom)(window.ethereum) : (0, viem_1.http)(),
            });
        }
    }
    async getAccount() {
        let account;
        if (this.privateKeyAccount) {
            account = this.privateKeyAccount;
        }
        else {
            const accounts = await this.walletClient.getAddresses();
            account = { address: accounts[0] };
        }
        return account;
    }
    async signTypedData({ message, types, primaryType, }) {
        const data = {
            domain: {
                name: 'sign.global',
                version: '1',
            },
            message: message,
            primaryType,
            types: {
                EIP712Domain: [
                    { name: 'name', type: 'string' },
                    { name: 'version', type: 'string' },
                ],
                ...types,
            },
        };
        const account = await this.getAccount();
        const signTypedData = this.privateKeyAccount
            ? account.signTypedData
            : this.walletClient.signTypedData;
        const signature = await signTypedData({
            account: account.address,
            ...data,
        });
        return {
            message: data,
            signature,
            publicKey: account.address,
        };
    }
    async signMessage(message) {
        const account = await this.getAccount();
        const signMessage = this.privateKeyAccount
            ? account.signMessage
            : this.walletClient.signMessage;
        const signature = await signMessage({
            account: account.address,
            message,
        });
        return {
            signature,
            message,
            publicKey: account.address,
        };
    }
}
exports.EVMOffChainClient = EVMOffChainClient;
